numpy/f2py/cfuncs.py

```diff
@@ -623,71 +646,119 @@
 }"""
 needs['try_pyarr_from_string'] = ['STRINGCOPYN', 'PRINTPYOBJERR', 'string']
 cfuncs['try_pyarr_from_string'] = """\
-static int try_pyarr_from_string(PyObject *obj,const string str) {
+/*
+  try_pyarr_from_string copies str[:len(obj)] to the data of an `ndarray`.
+
+  if the specified len==-1, str must be null-terminated.
+*/
+static int try_pyarr_from_string(PyObject *obj,
+                                 const string str, const int len) {
+#ifdef DEBUGCFUNCS
+fprintf(stderr, "try_pyarr_from_string(str='%s', len=%d, obj=%p)\\n",
+        (char*)str,len, obj);
+#endif
     PyArrayObject *arr = NULL;
-    if (PyArray_Check(obj) && (!((arr = (PyArrayObject *)obj) == NULL)))
-        { STRINGCOPYN(PyArray_DATA(arr),str,PyArray_NBYTES(arr)); }
-    return 1;
+    if (PyArray_Check(obj) && (!((arr = (PyArrayObject *)obj) == NULL))) {
+        string buf = PyArray_DATA(arr);
+        npy_intp n = len;
+        if (n == -1) {
+            /* Assuming null-terminated str. */
+            n = strlen(str);
+        }
+        if (n > PyArray_NBYTES(arr)) {
+            n = PyArray_NBYTES(arr);
+        }
```

Comment:
This check is enough; the array `np.arange(4, dtype=np.uint8)[::-1]` has a `data` pointer that points to the `3`, and writing four bytes from that point will result in three of them being written out of range.

Reply:
`try_pyarr_from_string` is used only for `intent(inout)` arguments. Such arguments must be contiguous. Hence, at this point, we can rely on not having negative strides. Added the following line to the function doc:
```
  If obj is ndarray, it is assumed to be C-contiguous.
```

Reply:
Can you add an `assert` that the array is contiguous just to be on the safe side?

Reply:
(or perhaps throw an error, as we do if `(!ISCONTIGUOUS(arr))` in some other code visible in this PR

Reply:
Handling of non-contiguous array input to `intent(inout)` string argument is implemented in `string_from_pyobj`, see
https://github.com/numpy/numpy/blob/3dcd9248d1f77cd2b0f97a5e9adb8e3cdcb3421f/numpy/f2py/cfuncs.py#L654-L660

The usage of all relevant functions can be seen here:
https://github.com/numpy/numpy/blob/3dcd9248d1f77cd2b0f97a5e9adb8e3cdcb3421f/numpy/f2py/rules.py#L949-L971

Hence there is no need to duplicate checking invalid user inputs in `try_pyarr_from_string`. Moreover, `try_pyarr_from_string` is not an user-facing function.

Reply:
That's a convincing argument for not bothering to check and throw an exception, but I think it's still worth adding a `assert(ISCONTIGUOUS(arr))` to ensure that no one changes the contract of that function by accident in future.

Reply:
Added the assert statement.